AX: Incomplete object information with focus event on dynamic created div

pgorszkowski-igalia opened the pull request:
#### 114392b51f9c322d303c0e44142d53d89276f66b
<pre>
AX: Incomplete object information with focus event on dynamic created div
<a href="https://bugs.webkit.org/show_bug.cgi?id=270433">https://bugs.webkit.org/show_bug.cgi?id=270433</a>

Reviewed by Tyler Wilcock.

When the element is dynamically created and focused, the focus change event should be
deferred in case renderer is not available yet but document needs style recalculation.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::shouldDeferFocusChange):
(WebCore::AXObjectCache::onFocusChange):
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestWebKitAccessibility.cpp:
(testAccessibleStateChangedFocus):
(beforeAll):
* LayoutTests/accessibility/focus-new-element.html: Added.
* LayoutTests/platform/glib/accessibility/focus-new-element-expected.txt: Added.
* LayoutTests/platform/ios-simulator-wk2/accessibility/focus-new-element-expected.txt: Added.
* LayoutTests/platform/mac/accessibility/focus-new-element-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/275980@main">https://commits.webkit.org/275980@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c3a90b9d8811257a7d02b4682327073f764fbf5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43411 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45822 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46046 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39537 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45715 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26216 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19860 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35886 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19472 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37390 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16863 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17048 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38452 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1467 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39606 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38763 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47590 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18358 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15076 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42670 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19863 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/37682 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41334 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9667 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20043 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19494 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1672,9 +1672,25 @@ void AXObjectCache::onPageActivityStateChange(OptionSet<ActivityState> newState)
+
+    auto* renderer = node->renderer();
+    if (renderer && (renderer->beingDestroyed() || renderer->needsLayout()))
+        return false;
```
Hmm, why is it correct to not defer focus change handling for renderers that need layout, or when we're in a render tree update / layout? That feels opposite to what I'd expect. Historically, updating the accessibility tree when layout is dirty has led to incorrect tree updates and crashes. But maybe this case is different for some reason?

pgorszkowski-igalia replied:
You are right, it is my fault. Previously, focus change was deferred if renderer needs layout. I will change this implementation. 
This mistake was not detected by any test, which is strange. Probably it is difficult to write a test for such case. I will analyze it and try to write also test for that.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1715,9 +1715,26 @@ void AXObjectCache::onPageActivityStateChange(OptionSet<ActivityState> newState)
 void AXObjectCache::onFocusChange(Node* oldNode, Node* newNode)
 {
-    if (nodeAndRendererAreValid(newNode) && rendererNeedsDeferredUpdate(*newNode->renderer())) {
+    if (shouldDeferFocusNodeChange(newNode)) {
```
`shouldDeferFocusNodeChange` now seems very similar to `rendererNeedsDeferredUpdate`. Why does the former need to be added? Also, why remove the `nodeAndRendererAreValid` check?

pgorszkowski-igalia replied:
onFocusChange can be called before newNode has a renderer but document.needsStyleRecalc() and that is why we need to defer it.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1734,7 +1751,7 @@ void AXObjectCache::onFocusChange(Node* oldNode, Node* newNode)
             m_deferredFocusedNodeChange = { oldNode, newNode };
 
-        if (!newNode->renderer()->needsLayout() && !m_performCacheUpdateTimer.isActive())
+        if (!((newNode->renderer() && newNode->renderer()->needsLayout()) || m_performCacheUpdateTimer.isActive()))
```
`|| m_performCacheUpdateTimer.isActive()` this doesn't make sense to me. We should avoid starting the timer if it's already active, which the code prior to this PR did, and now we do the opposite.

pgorszkowski-igalia replied:
The logic is the same as before, so the timer will not be started if it is active:
`if (!A && !B)` is equal to `if (!(A || B))` 

The only change here is that renderer can be null (that is why I add shouldDeferFocusNodeChange) so I need to check it.

twilco commented on Tools/TestWebKitAPI/Tests/WebKitGtk/TestWebKitAccessibility.cpp:
```diff
@@ -641,6 +641,40 @@ static void testAccessibleState(AccessibilityTest* test, gconstpointer)
     g_assert_true(atspi_state_set_contains(stateSet.get(), ATSPI_STATE_SHOWING));
 }
 
+static void testAccessibleStateChangedFocus(AccessibilityTest* test, gconstpointer)
+{
+    test->showInWindow(800, 600);
+    test->loadHtml(
+        "<html>"
+        "  <body>"
+        "    <div id='container'></div>"
+        "  </body>"
+        "</html>",
+        nullptr);
+    test->waitUntilLoadFinished();
+
+    auto testApp = test->findTestApplication();
+    g_assert_true(ATSPI_IS_ACCESSIBLE(testApp.get()));
+
+    auto documentWeb = test->findDocumentWeb(testApp.get());
+    g_assert_true(ATSPI_IS_ACCESSIBLE(documentWeb.get()));
+
+    test->startEventMonitor(nullptr, { "object:state-changed" });
+    test->runJavaScriptAndWaitUntilFinished(
+        "container=document.getElementById('container');"
+        "var i = document.createElement('div');"
+        "i.setAttribute('tabindex', '-1');"
+        "container.appendChild(i);"
+        "i.innerHTML = 'TEST';"
+        "i.focus();",
+        nullptr);
+    auto events = test->stopEventMonitor(1);
+    g_assert_cmpuint(events.size(), ==, 1);
+    g_assert_cmpstr(events[0]->type, ==, "object:state-changed:focused");
+    auto* div = events[0]->source;
+    g_assert_true(ATSPI_IS_ACCESSIBLE(div));
+}
```
It feels unfortunate that the bug you're fixing is in WebCore (vs. the platform layer), but the test being added is GTK only. If this is a bug in WebCore, then the bad / incomplete state situation you're fixing should be observable from a layout test, using the platform-agnostic AccessibilityUIElement APIs.

Also, what is the incomplete / bad state that manifests when this bug happens? Is it that the object is erroneously ignored (a.k.a `!ATSPI_IS_ACCESSIBLE(div)`)? Or something else?

pgorszkowski-igalia replied:
You are right, it should be possible to create a Layout test for that. I will try to do that.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1714,12 +1714,30 @@ void AXObjectCache::onPageActivityStateChange(OptionSet<ActivityState> newState)
 }
 
-void AXObjectCache::onFocusChange(Node* oldNode, Node* newNode)
+static bool shouldDeferFocusNodeChange(Element* element)
```
Sorry, I should've noticed this earlier, but IMO this function name could be a bit more concise, e.g. `shouldDeferFocusChange`.

twilco reviewed (approved):
Thanks for fixing this!

pgorszkowski-igalia committed:
AX: Incomplete object information with focus event on dynamic created div
https://bugs.webkit.org/show_bug.cgi?id=270433

Reviewed by Tyler Wilcock.

When the element is dynamically created and focused, the focus change event should be
deferred in case renderer is not available yet but document needs style recalculation.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::shouldDeferFocusChange):
(WebCore::AXObjectCache::onFocusChange):
* Tools/TestWebKitAPI/Tests/WebKitGtk/TestWebKitAccessibility.cpp:
(testAccessibleStateChangedFocus):
(beforeAll):
* LayoutTests/accessibility/focus-new-element.html: Added.
* LayoutTests/platform/glib/accessibility/focus-new-element-expected.txt: Added.
* LayoutTests/platform/ios-simulator-wk2/accessibility/focus-new-element-expected.txt: Added.
* LayoutTests/platform/mac/accessibility/focus-new-element-expected.txt: Added.

Canonical link: https://commits.webkit.org/275980@main
